### PR TITLE
webhook for twilio (no more twilio functions)

### PR DESCRIPTION
**Pull request by pdsarin**
Please review carefully given all the nonsense with From/To.

Can't go live in prod until we have idempotence figured out.

### CODE REVIEW

File: aws_lambdas/twilio_webhook.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+def handler(event, context):
+    kinesis = boto3.client("kinesis")
+    stage = os.getenv("STAGE")
+
+    validator = RequestValidator(os.getenv("TWILIO_AUTH_TOKEN"))
+    url = f"https://{event['headers']['Host']}/{stage}{event['path']}"
+    form = {
+        split_pair[0]: unquote_plus(split_pair[1])
+        for split_pair in [kvpair.split("=") for kvpair in event["body"].split("&")]
+    }
```
Comment (pdsarin) [inline aws_lambdas/twilio_webhook.py:24]: yeah, this is gross, but it gets a slightly different result than `urllib.parse_qs` - and that results in signature validation failure.

File: aws_lambdas/twilio_webhook.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        return {"statusCode": 403}
+    if "MessageStatus" in form:
+        kinesis.put_record(
+            Data=json.dumps({"type": "STATUS_UPDATE", "payload": form}),
+            PartitionKey=event["To"],
+            StreamName=f"message-log-{stage}",
+        )
+        return
+
+    CommandPublisher().publish_process_sms_command(form["From"], form["Body"])
```
Comment (pdsarin) [inline aws_lambdas/twilio_webhook.py:37]: forgot to log the inbound message

File: aws_lambdas/twilio_webhook.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    form = extract_form(event)
+    if not is_signature_valid(event, form, stage):
+        logging.warning("signature validation failed")
+        return {"statusCode": 403}
+    if "MessageStatus" in form:
+        logging.info(f"Outbound message to {form['To']}: Recording STATUS_UPDATE in message log")
+        kinesis.put_record(
+            Data=json.dumps({"type": "STATUS_UPDATE", "payload": form}),
+            PartitionKey=form["To"],
```
Comment (jeffreysilver) [inline aws_lambdas/twilio_webhook.py:28]: For my own sanity:

- Message status events are published after twilio receives a callback after sending an outbound message
- Outbound message means that our system is sending a message TO a user
- We should be using the "To" phone number
- This is correct

File: aws_lambdas/twilio_webhook.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            StreamName=f"message-log-{stage}",
+        )
+        return
+
+    logging.info(f"Inbound message from {form['From']}: Recording INBOUND_SMS in message log")
+    CommandPublisher().publish_process_sms_command(form["From"], form["Body"])
+    kinesis.put_record(
+        Data=json.dumps({"type": "INBOUND_SMS", "payload": form}),
+        PartitionKey=form["From"],
```
Comment (jeffreysilver) [inline aws_lambdas/twilio_webhook.py:37]: For my own sanity:

- Inbound sms events are published after twilio receives a message from a user 
- Inbound message means that our system is receiving a message From a user
- We should be using the "From" phone number
- This is correct

**Pull request merged**